File: 123.买卖股票的最佳时机-iii.c
/*
 * @lc app=leetcode.cn id=123 lang=c
 *
 * [123] 买卖股票的最佳时机 III
 */

// @lc code=start

int maxProfit(int *prices, int pricesSize)
{
  int i, **dp;
  dp = (int **)malloc(sizeof(int *) * pricesSize);
  for (i = 0; i < pricesSize; i++)
    dp[i] = (int *)malloc(sizeof(int) * 5);
  dp[0][0] = 0;
  dp[0][1] = -prices[0];
  dp[0][2] = 0;
  dp[0][3] = -prices[0];
  dp[0][4] = 0;

  for (i = 1; i < pricesSize; i++)
  {
    dp[i][0] = dp[i - 1][0];
    dp[i][1] = dp[i - 1][1] > dp[i - 1][0] - prices[i] ? dp[i - 1][1] : dp[i - 1][0] - prices[i];
    dp[i][2] = dp[i - 1][2] > dp[i - 1][1] + prices[i] ? dp[i - 1][2] : dp[i - 1][1] + prices[i];
    dp[i][3] = dp[i - 1][3] > dp[i - 1][2] - prices[i] ? dp[i - 1][3] : dp[i - 1][2] - prices[i];
    dp[i][4] = dp[i - 1][4] > dp[i - 1][3] + prices[i] ? dp[i - 1][4] : dp[i - 1][3] + prices[i];
  }

  return dp[pricesSize - 1][4];
}
// @lc code=end

File: store/123.买卖股票的最佳时机-iii.js
/*
 * @lc app=leetcode.cn id=123 lang=javascript
 *
 * [123] 买卖股票的最佳时机 III
 */

// @lc code=start
/**
 * @param {number[]} prices
 * @return {number}
 */
var maxProfit = function (prices) {
  let dp = [-prices[0], 0, -prices[0], 0];

  for (let i = 1; i < prices.length; i++) {
    dp[0] = Math.max(dp[0], -prices[i]);
    dp[1] = Math.max(dp[1], dp[0] + prices[i]);
    dp[2] = Math.max(dp[2], dp[1] - prices[i]);
    dp[3] = Math.max(dp[3], dp[2] + prices[i]);
  }

  return dp[3];
};
// @lc code=end
